### PR TITLE
[srvls][SRVOCF-278] Adding Nodejs functions docs

### PR DESCRIPTION
**Pull request by abrennan89**
https://deploy-preview-32551--osdocs.netlify.app/openshift-enterprise/latest/serverless/functions/serverless-developing-nodejs-functions.html

### CODE REVIEW

File: modules/serverless-nodejs-context-object-reference.adoc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+.Example command
+[source,terminal]
+----
+curl http://example.com?name=tiger
+----
+
+.Example output
+[source,terminal]
+----
+{"level":30,"time":1604511655265,"pid":3430203,"hostname":"localhost.localdomain","reqId":1,"msg":"tiger"}
```
Comment (lance) [inline modules/serverless-nodejs-context-object-reference.adoc:64]: There should only be one log line.

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+In a Knative system, if a function deployed as a service is invoked by an event broker sending a CloudEvent, the broker will examine the response. If the response is a CloudEvent, this event will then be handled by the broker.
+
+.Example context object method
+[source,js]
+----
+// Expects to receive a CloudEvent with customer data
+function processCustomer(context, customer) {
+  // process the customer
+  const processed = processCustomer(customer);
+  return context.cloudEventResponse(customer);
```
Comment (lance) [inline modules/serverless-nodejs-functions-context-objects.adoc:32]:
```suggestion
  return context.cloudEventResponse(customer)
    .source('/processCustomer')
    .type('fn.process.customer')
    .response();
```

This is actually a better example, showing customization of the source and type of return event.

File: modules/serverless-nodejs-function-return-values.adoc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+= Node.js function return values
+
+Functions can return any valid JavaScript type or can have no return value. When a function has no return value specified, and no failure is indicated, the caller will receive a `204 No Content` response.
+
+Functions may also return a CloudEvent, or a `Message` object in order to push events into the Knative Eventing system. In this case, the developer is not required to understand or implement the CloudEvent messaging specification. Headers and other relevant information from the returned values are extracted and sent with the response.
+
+.Example
+[source,javascript]
+----
+function processCustomer(context, customer) {
```
Comment (lance) [inline modules/serverless-nodejs-function-return-values.adoc:15]:
```suggestion
function handle(context, customer) {
```

File: modules/serverless-nodejs-function-return-values.adoc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+[id="serverless-nodejs-function-return-values-headers_{context}"]
+== Returning headers
+
+You can set a response header by adding a `headers` property to the `return` object. These headers will be extracted and sent with the response to the caller.
+
+.Example response header
+[source,javascript]
+----
+function processCustomer(context, customer) {
```
Comment (lance) [inline modules/serverless-nodejs-function-return-values.adoc:32]:
```suggestion
function handle(context, customer) {
```

File: modules/serverless-nodejs-function-return-values.adoc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+[id="serverless-nodejs-function-return-values-status-codes_{context}"]
+== Returning status codes
+
+You can set a status code that is returned to the caller by adding a `statusCode` property to the `return` object:
+
+.Example status code
+[source,javascript]
+----
+function processCustomer(context, customer) {
```
Comment (lance) [inline modules/serverless-nodejs-function-return-values.adoc:47]:
```suggestion
function handle(context, customer) {
```

File: modules/serverless-nodejs-function-return-values.adoc [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  }
+}
+----
+
+Status codes can also be set for errors that are created and thrown by the function:
+
+.Example error status code
+[source,javascript]
+----
+function processCustomer(context, customer) {
```
Comment (lance) [inline modules/serverless-nodejs-function-return-values.adoc:60]:
```suggestion
function handle(context, customer) {
```

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+[id="serverless-nodejs-functions-context-objects_{context}"]
+= Node.js context objects
+
+Functions are invoked by providing a `context` object as the first parameter.
+
+.Example context object
+[source,js]
+----
+function processPurchase(context, data)
```
Comment (lance) [inline modules/serverless-nodejs-functions-context-objects.adoc:13]:
```suggestion
function handle(context, data)
```

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+The `context` object has a single method, `cloudEventResponse()` that accepts a data value and returns a CloudEvent.
+
+In a Knative system, if a function deployed as a service is invoked by an event broker sending a CloudEvent, the broker will examine the response. If the response is a CloudEvent, this event will then be handled by the broker.
+
+.Example context object method
+[source,js]
+----
+// Expects to receive a CloudEvent with customer data
+function processCustomer(context, customer) {
```
Comment (lance) [inline modules/serverless-nodejs-functions-context-objects.adoc:29]:
```suggestion
function handle(context, customer) {
```

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  "productId": "6543210"
+}
+----
+
+When invoked, the second parameter to the function, after the `context` object, will be a JavaScript object that has `customerId` and `productId` properties.
+
+.Example signature
+[source,javascript]
+----
+function processPurchase(context, data)
```
Comment (lance) [inline modules/serverless-nodejs-functions-context-objects.adoc:54]:
```suggestion
function handle(context, data)
```

**Review by lance — COMMENTED** [review]
Some small suggestions. Overall, I have received feedback that including random domain information in the examples is confusing. For example, a function named `processPurchase()` may not make any sense to someone writing functions for IoT. So, I  have suggested renaming all of these example functions to simply `handle()`.

File: modules/serverless-nodejs-context-object-reference.adoc [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+[id="serverless-nodejs-context-object-reference-log_{context}"]
+== log
+
+Provides a logging object that can be used to write output to the cluster logs. The log adheres to the Pino logging API (https://getpino.io/#/docs/api).
+
+.Example
+[source,js]
+----
+Function myFunction(context) {
```
Comment (lance) [inline modules/serverless-nodejs-context-object-reference.adoc:18]:
```suggestion
function handle(context) {
```
I think, similar to the "return values" adoc, all of our examples should just be `function handle()`. Note the lower case `f` in `function`. Sorry for the last minute recommendation here.

File: modules/serverless-nodejs-context-object-reference.adoc [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+[id="serverless-nodejs-context-object-reference-query_{context}"]
+== query
+
+Returns the query string for the request, if any, as key value pairs. These attributes are also found on the context object itself.
+
+.Example query
+[source,js]
+----
+Function myFunction(context) {
```
Comment (lance) [inline modules/serverless-nodejs-context-object-reference.adoc:45]:
```suggestion
function handle(context) {
```

File: modules/serverless-nodejs-context-object-reference.adoc [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+[id="serverless-nodejs-context-object-reference-body_{context}"]
+== body
+
+Returns the request body if any. If the request body contains JSON code, this will be parsed so that the attributes are directly available.
+
+.Example
+[source,js]
+----
+Function myFunction(context) {
```
Comment (lance) [inline modules/serverless-nodejs-context-object-reference.adoc:75]:
```suggestion
function handle(context) {
```

File: modules/serverless-nodejs-context-object-reference.adoc [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+[id="serverless-nodejs-context-object-reference-headers_{context}"]
+== headers
+
+Returns the HTTP request headers as an object.
+
+.Example header
+[source,js]
+----
+Function myFunction(context) {
```
Comment (lance) [inline modules/serverless-nodejs-context-object-reference.adoc:103]:
```suggestion
function handle(context) {
```

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+[id="serverless-nodejs-functions-context-objects_{context}"]
+= Node.js context objects
+
+Functions are invoked by providing a `context` object as the first parameter.
+
+.Example context object
+[source,js]
+----
+function processPurchase(context, data)
```
Comment (lance) [inline modules/serverless-nodejs-functions-context-objects.adoc:13]:
```suggestion
function handle(context, data)
```

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  "productId": "6543210"
+}
+----
+
+When invoked, the second parameter to the function, after the `context` object, will be a JavaScript object that has `customerId` and `productId` properties.
+
+.Example signature
+[source,javascript]
+----
+function processPurchase(context, data)
```
Comment (lance) [inline modules/serverless-nodejs-functions-context-objects.adoc:57]:
```suggestion
function handle(context, data)
```

File: modules/serverless-testing-nodejs-functions.adoc [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+Node.js functions can be tested locally on your computer. In the default project that is created when you create a function using `kn func create`, there is a *test* folder that contains some simple unit and integration tests.
+
+.Procedure
+
+* Run the tests:
++
+[source,terminal]
+----
+$ go test
```
Comment (lance) [inline modules/serverless-testing-nodejs-functions.adoc:16]:
```suggestion
$ npm test
```

**Review by lance — COMMENTED** [review]
@abrennan89 - a few recommended changes on function names, as before. I should have caught these the first time!

**Comment by jrangelramos** [comment]
It looks good to me!
QE Approved

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+----
+
+This object provides access to the incoming HTTP request information, including the HTTP request method, any query strings or headers sent with the request, the HTTP version, and the request body. Incoming requests that contain a CloudEvent attach the incoming instance of the CloudEvent to the context object, so that it can be accessed by using `context.cloudevent`.
+
+[id="serverless-nodejs-functions-context-objects-methods_{context}"]
+== Context object methods
+
+The `context` object has a single method, `cloudEventResponse()` that accepts a data value and returns a CloudEvent.
+
+In a Knative system, if a function deployed as a service is invoked by an event broker sending a CloudEvent, the broker will examine the response. If the response is a CloudEvent, this event will then be handled by the broker.
```
Comment (lbarbeevargas) [inline modules/serverless-nodejs-functions-context-objects.adoc:23]: Some wording considerations to put this is present tense:

> In a Knative system, if a function deployed as a service is invoked by an event broker sending a CloudEvent, the broker examines the response. If the response is a CloudEvent, this event is handled by the broker.

File: modules/serverless-nodejs-functions-context-objects.adoc [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    .source('/handle')
+    .type('fn.process.customer')
+    .response();
+}
+----
+
+[id="serverless-nodejs-functions-context-objects-cloudevent-data_{context}"]
+== CloudEvent data
+
+If the incoming request is a CloudEvent, any data associated with the CloudEvent is extracted from the event and provided as a second parameter. For example, if a CloudEvent is received which contains a JSON string similar to the following in its data property:
```
Comment (lbarbeevargas) [inline modules/serverless-nodejs-functions-context-objects.adoc:42]: s/which contains/that contains/

And wording consideration for readability:

> For example, if a CloudEvent is received that contains a JSON string in its data property that is similar to the following:

File: modules/serverless-nodejs-function-return-values.adoc [hunk 19]
```diff
@@ -0,0 +1,68 @@
+// Module included in the following assemblies
+//
+// * /serverless/functions/serverless-developing-nodejs-functions.adoc
+
+[id="serverless-nodejs-function-return-values_{context}"]
+= Node.js function return values
+
+Functions can return any valid JavaScript type or can have no return value. When a function has no return value specified, and no failure is indicated, the caller will receive a `204 No Content` response.
```
Comment (lbarbeevargas) [inline modules/serverless-nodejs-function-return-values.adoc:8]: s/will receive/receives/

File: modules/serverless-nodejs-function-return-values.adoc [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+Functions may also return a CloudEvent, or a `Message` object in order to push events into the Knative Eventing system. In this case, the developer is not required to understand or implement the CloudEvent messaging specification. Headers and other relevant information from the returned values are extracted and sent with the response.
```
Comment (lbarbeevargas) [inline modules/serverless-nodejs-function-return-values.adoc:10]: s/may/can/
Rewording considerations for clarity if a CloudEvent is a `Message` object:

> Functions can also return a CloudEvent, or a `Message` object, in order to push events into the Knative Eventing system.

If a CloudEvent and a `Message` object are two different things:

> Functions can also return a CloudEvent or a `Message` object in order to push events into the Knative Eventing system.

File: modules/serverless-nodejs-function-return-values.adoc [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    source: 'customer.processor',
+    type: 'customer.processed'
+  })
+}
+----
+
+[id="serverless-nodejs-function-return-values-headers_{context}"]
+== Returning headers
+
+You can set a response header by adding a `headers` property to the `return` object. These headers will be extracted and sent with the response to the caller.
```
Comment (lbarbeevargas) [inline modules/serverless-nodejs-function-return-values.adoc:27]: s/will be extracted/are extracted/

**Review by lbarbeevargas — COMMENTED** [review]
This looks great! A few wording considerations and suggestions. Also, please check instances of `[source,js]` in the function examples. They should be `[source,javascript]`. Please let me know if you have any questions about my feedback.

File: modules/serverless-nodejs-context-object-reference.adoc [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+
+The `context` object has several properties that can be accessed by the function developer.
+
+[id="serverless-nodejs-context-object-reference-log_{context}"]
+== log
+
+Provides a logging object that can be used to write output to the cluster logs. The log adheres to the link:https://getpino.io/#/docs/api[Pino logging API].
+
+.Example log
+[source,js]
```
Comment (lbarbeevargas) [inline modules/serverless-nodejs-context-object-reference.adoc:16]: One more `[source,js]` -> `[source,javascript]` 👍

Comment (abrennan89) [reply modules/serverless-nodejs-context-object-reference.adoc:16]: Oops, thank you!

**Review by lbarbeevargas — COMMENTED** [review]
Updates LGTM.

**Pull request merged**